File: mntf/cavidade_fdm_explicit.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

static const double PI = 3.1415926535;
static const double TOLERANCE_PRESSURE = 1e-9;
static double overdx2, overdy2, over2dx, over2dy, overdtdx, overdtdy, dtover2dx, dtover2dy, mu, dt;

double top_speed(const double x)
{
    return pow(sin(PI * x), 2);
}

double initial_horizontal_speed(const double x, const double y)
{
    return pow(sin(PI * x), 2) * y * (3 * y - 2);
}

double initial_vertical_speed(const double x, const double y)
{
    return -2 * sin(PI * x) * cos(PI * x) * y * y * (y - 1);
}

void init_horizontal_speed_matrix(const int nx, const int ny, double u[nx][ny + 1])
{
    int i, j;
    double dx, dy;
    dx = 1.0 / (nx - 1);
    dy = 1.0 / (ny - 1);
    for (i = 0; i < nx; i++)
    {
        for (j = 1; j < ny; j++)
            u[i][j] = initial_horizontal_speed(i * dx, (j + 0.5) * dy);
        u[i][0] = -u[i][1];
        u[i][ny] = 2 * top_speed(i * dx) - u[i][ny - 1];
    }
}
void init_vertical_speed_matrix(const int nx, const int ny, double v[nx + 1][ny])
{
    int i, j;
    double dx, dy;
    dx = 1.0 / (nx - 1);
    dy = 1.0 / (ny - 1);
    for (j = 0; j < ny; j++)
    {
        for (i = 1; i < nx; i++)
            v[i][j] = initial_vertical_speed((i + 0.5) * dx, j * dy);
        v[0][j] = -v[1][j];
        v[nx][j] = -v[nx - 1][j];
    }
}

void print_matrix(const int nx, const int ny, const double matriz[nx][ny])
{
    int i, j;
    printf("\n");
    for (j = ny - 1; j >= 0; j--)
    {
        for (i = 0; i < nx; i++)
            printf("%c%.02f ", matriz[i][j] < 0 ? '-' : ' ', fabs(matriz[i][j]));
        printf("\n");
    }
}

void init_matrix_random_values(const int nx, const int ny, double matriz[nx][ny])
{
    int i, j;
    for (i = 0; i < nx; i++)
        for (j = 0; j < ny; j++)
            matriz[i][j] = rand();
}

int verify_sizes(const double dx, const double dy, const double dt, const double Re)
{
    if (dx < dt)
    {
        printf("Must have dt < dx: dx = %.1e; dt = %1.e", dx, dt);
        return 1;
    }
    if (dy < dt)
    {
        printf("Must have dt < dy: dx = %.1e; dt = %1.e", dy, dt);
        return 1;
    }
    if (1 < dx * dx * Re)
    {
        printf("The value of dx is too big! dx = %.1e > 1/sqrt(Re) = %.1e", dx, 1 / sqrt(Re));
        return 1;
    }
    if (1 < dy * dy * Re)
    {
        printf("The value of dy is too big! dy = %.1e > 1/sqrt(Re) = %.1e", dy, 1 / sqrt(Re));
        return 1;
    }
    if (0.25 * Re * dx * dx < dt)
    {
        printf("The value of dt is too big! dt = %.1e > 0.25*Re*dx^2 = %.1e", dt, 0.25 * Re * dx * dx);
        return 1;
    }
    if (0.25 * Re * dy * dy < dt)
    {
        printf("The value of dt is too big! dt = %.1e > 0.25*Re*dy^2 = %.1e", dt, 0.25 * Re * dy * dy);
        return 1;
    }
    return 0;
}

void update_ustar(const int nx, const int ny, const double u[nx][ny + 1], const double v[nx + 1][ny], double ustar[nx][ny + 1])
{
    int i, j;
    double meanv, laplace, dudx, dudy, R, dx;
    dx = 1.0 / (nx - 1);
    for (i = 1; i < nx - 1; i++)
        for (j = 1; j < ny; j++)
        {
            meanv = 0.25 * (v[i][j] + v[i - 1][j - 1] + v[i][j] + v[i - 1][j - 1]);
            dudx = over2dx * (u[i + 1][j] - u[i - 1][j]);
            dudy = over2dy * (u[i][j + 1] - u[i][j - 1]);
            laplace = overdx2 * (u[i + 1][j] - 2 * u[i][j] + u[i - 1][j]);
            laplace += overdy2 * (u[i][j + 1] - 2 * u[i][j] + u[i][j - 1]);
            R = mu * laplace - u[i][j] * dudx - meanv * dudy;
            ustar[i][j] = u[i][j] + dt * R;
        }
    for (j = 1; j < ny; j++)
    {
        ustar[0][j] = 0;
        ustar[nx - 1][j] = 0;
    }
    for (i = 0; i < nx + 1; i++)
    {
        ustar[i][0] = -ustar[i][1];
        ustar[i][ny] = 2 * top_speed(i * dx) - ustar[i][ny - 1];
    }
}

void update_vstar(const int nx, const int ny, const double u[nx][ny + 1], const double v[nx + 1][ny], double vstar[nx + 1][ny])
{
    int i, j;
    double meanu, laplace, dvdx, dvdy, R;
    for (i = 1; i < nx; i++)
        for (j = 1; j < ny - 1; j++)
        {
            meanu = 0.25 * (u[i][j] + u[i][j - 1] + u[i - 1][j] + u[i - 1][j - 1]);
            dvdx = over2dx * (v[i + 1][j] - v[i - 1][j]);
            dvdy = over2dy * (v[i][j + 1] - v[i][j - 1]);
            laplace = overdx2 * (v[i + 1][j] - 2 * v[i][j] + u[i - 1][j]);
            laplace += overdy2 * (v[i][j + 1] - 2 * v[i][j] + u[i][j - 1]);
            R = mu * laplace - v[i][j] * dvdy - meanu * dvdx;
            vstar[i][j] = v[i][j] + dt * R;
        }
    for (i = 1; i < nx; i++)
    {
        vstar[i][0] = 0;
        vstar[i][ny - 1] = 0;
    }
    for (j = 0; j < ny; j++)
    {
        vstar[0][j] = -vstar[1][j];
        vstar[nx][j] = -vstar[nx - 1][j];
    }
}

void update_force_pressure(const int nx, const int ny, const double ustar[nx][ny + 1], const double vstar[nx + 1][ny], double force_pressure[nx][ny])
{
    int i, j;
    for (i = 1; i < nx; i++)
        for (j = 1; j < ny; j++)
        {
            if (fabs(ustar[i + 1][j]) > 1)
                printf("ustar[i+1][j] = ustar[%d][%d] = %.3e\n", i + 1, j, ustar[i + 1][j]);
            if (fabs(vstar[i][j + 1]) > 1)
                printf("vstar[i][j+1] = vstar[%d][%d] = %.3e\n", i, j + 1, vstar[i][j + 1]);
            force_pressure[i][j] = overdtdx * (ustar[i + 1][j] - ustar[i][j]);
            force_pressure[i][j] += overdtdy * (vstar[i][j + 1] - vstar[i][j]);
        }
}

void update_pressure(const int nx, const int ny, const double force_pressure[nx][ny], double pressure[nx + 1][ny + 1])
{
    int i, j;
    double R, Rmax, lambda;
    do
    {
        Rmax = 0;
        for (i = 1; i < nx; i++)
        {
            for (j = 1; j < ny; j++)
            {
                R = force_pressure[i][j];
                lambda = 0;
                if (i > 1)
                {
                    R += overdx2 * (pressure[i][j] - pressure[i - 1][j]);
                    lambda -= overdx2;
                }
                if (i < nx - 1)
                {
                    R += overdx2 * (pressure[i][j] - pressure[i + 1][j]);
                    lambda -= overdx2;
                }
                if (j > 1)
                {
                    R += overdy2 * (pressure[i][j] - pressure[i][j - 1]);
                    lambda -= overdy2;
                }
                if (j < ny - 1)
                {
                    R += overdy2 * (pressure[i][j] - pressure[i][j + 1]);
                    lambda -= overdy2;
                }
                R /= lambda;
                pressure[i][j] += R;
                Rmax = fabs(R) > Rmax ? fabs(R) : Rmax;
            }
        }
    } while (Rmax > TOLERANCE_PRESSURE);
}

int main()
{
    srand(time(0));
    int nx = 4, ny = 5, nt = 10001;
    double dx, dy;
    float tmax = 3;
    double Re = 1;
    double uspeed[nx][ny + 1];
    double vspeed[nx + 1][ny];
    double pressure[nx + 1][ny + 1];
    double ustar[nx][ny + 1];
    double vstar[nx + 1][ny];
    double force_pressure[nx][ny];

    dx = 1.0 / (nx - 1);
    dy = 1.0 / (ny - 1);
    dt = tmax / (nt - 1);

    if (verify_sizes(dx, dy, dt, Re))
        return 1;

    init_matrix_random_values(nx, ny + 1, uspeed);
    init_matrix_random_values(nx + 1, ny, vspeed);
    init_matrix_random_values(nx, ny + 1, ustar);
    init_matrix_random_values(nx + 1, ny, vstar);
    init_matrix_random_values(nx + 1, ny + 1, pressure);
    init_matrix_random_values(nx, ny, force_pressure);

    init_horizontal_speed_matrix(nx, ny, uspeed);
    printf("Horizontal speed\n");
    print_matrix(nx, ny + 1, uspeed);
    init_vertical_speed_matrix(nx, ny, vspeed);
    printf("Vertical speed\n");
    print_matrix(nx + 1, ny, vspeed);

    overdx2 = 1.0 / (dx * dx);
    overdy2 = 1.0 / (dy * dy);
    over2dx = 1.0 / (2 * dx);
    over2dy = 1.0 / (2 * dy);
    overdtdx = 1.0 / (dt * dx);
    overdtdy = 1.0 / (dt * dy);
    dtover2dx = dt / (2 * dx);
    dtover2dy = dt / (2 * dy);
    mu = 1.0 / Re;

    /*for (k = 1; k < nt; k++)
{
}*/
    printf("Got here! Update Ustar!\n");
    update_ustar(nx, ny, uspeed, vspeed, ustar);
    printf("Printing U star\n");
    print_matrix(nx, ny + 1, ustar);
    printf("Got here! Update Vstar!\n");
    update_vstar(nx, ny, uspeed, vspeed, vstar);
    printf("Printing V star\n");
    print_matrix(nx + 1, ny, vstar);
    printf("Update force pressure\n");
    update_force_pressure(nx, ny, ustar, vstar, force_pressure);
    printf("Printing force pressure\n");
    print_matrix(nx, ny, force_pressure);
    printf("Update pressure\n");
    /* update_pressure(nx, ny, force_pressure, pressure); */
    printf("Printing pressure\n");
    print_matrix(nx, ny, force_pressure);
    return 0;
}

File: mntf/cavidade_fdm_explicit.py
import numpy as np
from matplotlib import pyplot as plt
from typing import Tuple
from tqdm import tqdm
from numba import jit
from compmec import nurbs
from ploter import plot_animated_field


def file_exists(filename: str) -> bool:
    try:
        file = open(filename, "r")
        file.close()
        return True
    except FileNotFoundError:
        return False

np.set_printoptions(precision=2, suppress=True)

nx, ny, nt = 31, 31, 10001
ntsave = 101
xmin, xmax = 0, 1
ymin, ymax = 0, 1
tmin, tmax = 0, 3
xmesh = np.linspace(xmin, xmax, nx)
ymesh = np.linspace(ymin, ymax, ny)
# tmesh = np.linspace(tmin, tmax, nt)
tsavemesh = np.linspace(tmin, tmax, ntsave)
dx, dy = 1/(nx-1), 1/(ny-1)
dt = (tmax-tmin)/(nt-1) 
Re = 100  # Reynolds

if dx < dt:
    raise ValueError("The value of dx is less than dt! dx = %.1e, dt = %.1e" % (dx, dt))
if 1/np.sqrt(Re) < dx:
    raise ValueError("The value of dx is too big! dx = %.1e, 1/sqrt(Re) = %.1e" % (dx, 1/np.sqrt(Re)))
if 0.25*Re*dx**2 < dt:
    raise ValueError("The value of dt is too big! dt = %.1e, Re*dx^2/4 = %.1e" % (dt, 0.25*Re*dx**2))


@np.vectorize
def U(x):  # Condicao no topo
    return np.sin(np.pi*x)**2

def verify_has_nan_inf(value: np.ndarray, name: str):
    if not np.isfinite(value).all():
        print(f"variavel {name}")
        print(value.T[::-1])
        raise ValueError(f"{name} constains np.nan or np.inf")

u = np.zeros((nx, ny+1), dtype="float64")
v = np.zeros((nx+1, ny), dtype="float64")
p = np.zeros((nx+1, ny+1), dtype="float64")
Uf = np.zeros((ntsave, nx, ny), dtype="float64")
Vf = np.zeros((ntsave, nx, ny), dtype="float64")
Pf = np.zeros((ntsave, nx, ny), dtype="float64")
ustar = np.copy(u)
vstar = np.copy(v)

divu = np.empty((nx, ny), dtype="float64")
divv = np.empty((nx, ny), dtype="float64")
meanu = np.empty((nx, ny), dtype="float64")
meanv = np.empty((nx, ny), dtype="float64")
Rvals = np.zeros((nx-1, ny-1), dtype="float64")  # Para calcular a pressao
matrix2D = np.empty((nx-1, ny-1), dtype="float64")  # Force to iterate computing pressure

overdx2 = 1/(dx**2)
overdy2 = 1/(dy**2)
overdtdx = 1/(dt*dx)
overdtdy = 1/(dt*dy)
dtover2dx = dt/(2*dx)
dtover2dy = dt/(2*dy)

def init_diagon2D(nx, ny, overdx2, overdy2):
    diagon2D = np.zeros((nx-1, ny-1), dtype="float64")  # Mesmo que o lambda
    diagon2Dx = np.zeros((nx-1, ny-1), dtype="float64")
    diagon2Dy = np.zeros((nx-1, ny-1), dtype="float64")
    diagon2D[:-1, :] -= overdx2
    diagon2D[1:, :] -= overdx2
    diagon2D[:, :-1] -= overdy2
    diagon2D[:, 1:] -= overdy2
    diagon2D[:, :] = 1/diagon2D[:, :]
    diagon2Dx[:, :] = overdx2*diagon2D[:, :]
    diagon2Dy[:, :] = overdy2*diagon2D[:, :]
    return diagon2D, diagon2Dx, diagon2Dy

# @jit(nopython=True)
def compute_ustar(u, divu, meanv, ustar, Uupper):
    divu[1:nx-1, 0:ny-1] = -2*(overdx2+overdy2)*u[1:nx-1, 0:ny-1]
    divu[1:nx-1, 0:ny-1] += overdx2*u[0:nx-2, 0:ny-1]
    divu[1:nx-1, 0:ny-1] += overdx2*u[2:nx, 0:ny-1]
    divu[1:nx-1, 0:ny-1] += overdy2*u[1:nx-1, 1:ny]
    divu[1:nx-1, 1:ny-1] += overdy2*u[1:nx-1, 0:ny-2]
    divu[1:nx-1, 0] += overdy2*u[1:nx-1,-1]
    meanv[1:nx-1, 0:ny-1] = v[1:nx-1, 0:ny-1]
    meanv[1:nx-1, 0:ny-1] += v[1:nx-1, 1:ny]
    meanv[1:nx-1, 0:ny-1] += v[0:nx-2, 0:ny-1]
    meanv[1:nx-1, 0:ny-1] += v[0:nx-2, 1:ny]
    meanv *= (1/4)
    # Comecamos o preenchimento
    ustar[1:nx-1, 0:ny-1] = u[1:nx-1, 0:ny-1]*(1 - dtover2dx*(u[2:nx, 0:ny-1] - u[1:nx-1, 0:ny-1]))
    ustar[1:nx-1, 0:ny-1] += (dt/Re)*divu[1:nx-1, 0:ny-1]
    ustar[1:nx-1, 0:ny-1] -= dtover2dy*meanv[1:nx-1, 0:ny-1]*(u[1:nx-1, 1:ny] - u[1:nx-1, 0:ny-1])
    # Condicoes de contorno
    ustar[0, 0:ny-1].fill(0)
    ustar[nx-1, 0:ny-1].fill(0)
    ustar[0:nx, -1] = -ustar[0:nx, 0]
    ustar[0:nx, ny-1] = 2*Uupper-ustar[0:nx, ny-2]

# @jit(nopython=True)
def compute_vstar(v, divv, meanu, vstar):
    divv[0:nx-1, 1:ny-1] = -2*(overdx2+overdy2)*v[0:nx-1, 1:ny-1]
    divv[0:nx-1, 1:ny-1] += overdy2*v[0:nx-1, 2:ny]
    divv[0:nx-1, 1:ny-1] += overdy2*v[0:nx-1, 0:ny-2]
    divv[0:nx-1, 1:ny-1] += overdx2*v[1:nx, 1:ny-1]
    divv[1:nx-1, 1:ny-1] += overdx2*v[0:nx-2, 1:ny-1]
    divv[0, 1:ny-1] += overdx2*v[-1,1:ny-1]
    meanu[0:nx-1, 1:ny-1] = u[0:nx-1, 1:ny-1]
    meanu[0:nx-1, 1:ny-1] += u[0:nx-1, 0:ny-2]
    meanu[0:nx-1, 1:ny-1] += u[1:nx, 1:ny-1]
    meanu[0:nx-1, 1:ny-1] += u[1:nx, 0:ny-2]
    meanu *= (1/4)
    # Comecamos o preenchimento
    vstar[0:nx-1, 1:ny-1] = v[0:nx-1, 1:ny-1]*(1-dtover2dy*(v[0:nx-1, 2:ny] - v[0:nx-1, 1:ny-1]))
    vstar[0:nx-1, 1:ny-1] += (dt/Re)*divv[0:nx-1, 1:ny-1]
    vstar[0:nx-1, 1:ny-1] -= dtover2dx*meanu[0:nx-1, 1:ny-1]*(v[1:nx, 1:ny-1] - v[0:nx-1, 1:ny-1])
    # Condicoes de contorno
    vstar[0:nx, 0].fill(0)
    vstar[0:nx, ny-1].fill(0)
    vstar[-1, 0:ny] = -vstar[0, 0:ny]
    vstar[nx-1, 0:ny] = -vstar[nx-2, 0:ny]

# @jit(nopython=True)
def compute_matrix2D(matrix2D, diagon2D):
    for i in range(0, nx-1):
        matrix2D[i, 0:ny-1] = overdtdy*(vstar[i, 1:ny]-vstar[i, 0:ny-1])
    for j in range(0, ny-1):
        matrix2D[0:nx-1, j] += overdtdx*(ustar[1:nx, j]-ustar[0:nx-1, j])
    # matrix2D *= diagon2D
    return matrix2D

# @jit(nopython=True)  
def compute_pressure(matrix2D, diagon2Dx, diagon2Dy, p):
    TOLERANCE = 1e-6
    ITERMAX = 200
    iter = 0
    while True:
        Rmax = 0
        for i in range(1, nx-1):
            for j in range(1, ny-1):
                lamda = 0
                R = overdtdx*(ustar[i+1, j]-ustar[i, j])
                R += overdtdy*(vstar[i, j+1]-vstar[i, j])
                if i == 0:
                    lamda -= overdx2
                    R -= overdx2*(p[i+1, j] - p[i, j])
                elif i == nx-2:
                    lamda -= overdx2
                    R -= overdx2*(p[i-1, j] - p[i, j])
                else:
                    lamda -= 2*overdx2
                    R -= overdx2*(p[i-1, j] - 2*p[i, j] + p[i+1, j])
                if j == 0:
                    lamda -= overdy2
                    R -= overdy2*(p[i, j+1] - p[i, j])
                elif j == ny-2:
                    lamda -= overdy2
                    R -= overdy2*(p[i, j-1] - p[i, j])
                else:
                    lamda -= 2*overdy2
                    R -= overdy2*(p[i, j-1] - 2*p[i, j] + p[i, j+1])
                R /= lamda
                p[i, j] += R
                if np.abs(R) > Rmax:
                    Rmax = np.abs(R)
        if Rmax < TOLERANCE:
            break
        iter += 1
        if iter == ITERMAX:
            error = "Maximum iteration " + str(ITERMAX) + " reached to compute pressure"
            raise ValueError(error)
    for i in range(nx-1):
        p[i, -1] = p[i, 0]
    p[0:nx-1, ny-1] = p[0:nx-1, ny-2]
    p[-1, :] = p[0, :]
    p[nx-1, :] = p[nx-2, :]
    return p

# @jit(nopython=True)
def compute_newu(ustar, p, u):
    for j in range(-1, ny):
        u[1:nx-1, j] = ustar[1:nx-1, j] - (dt/dx)*(p[1:nx-1, j]-p[0:nx-2, j])

# @jit(nopython=True)
def compute_newv(vstar, p, v):
    for i in range(-1, nx):
        v[i, 1:ny-1] = vstar[i, 1:ny-1] - (dt/dy)*(p[i, 1:ny-1]-p[i, 0:ny-2])



if __name__ == "__main__":
    filename = "cavidade-fdm-explicit-Re%.1e-nx%dny%dnt%dtmax%.1edt%.1e"%(Re, nx, ny, ntsave, tmax, dt)
    compute_new = True
    if file_exists(filename):
        compute_new = False
        arrays = np.load(filename)
        if not np.allclose(xmesh, arrays["xmesh"]):
            compute_new = True
        if not np.allclose(ymesh, arrays["ymesh"]):
            compute_new = True
        if not np.allclose(tsavemesh, arrays["tmesh"]):
            compute_new = True
        if not compute_new:
            Uf[:, :, :] = arrays["U"]
            Vf[:, :, :] = arrays["V"]
            Pf[:, :, :] = arrays["P"]
        
    if compute_new:   
        diagon2D, diagon2Dx, diagon2Dy = init_diagon2D(nx, ny, overdx2, overdy2)
        u.fill(0)
        v.fill(0)
        p.fill(0)

        # Apenas para compilar e nao tomar tempo:
        compute_ustar(u, divu, meanv, ustar, np.zeros(xmesh.shape))
        compute_vstar(v, divv, meanu, vstar)
        matrix2D = compute_matrix2D(matrix2D, diagon2D)
        p = compute_pressure(matrix2D, diagon2Dx, diagon2Dy, p)
        compute_newu(ustar, p, u)
        compute_newv(vstar, p, v)

        # Agora vamos ao calculo
        Uupper = 2*U(xmesh)
        u[:,ny-1] = Uupper[:]  # Boundary condition
        try:
            Uf[0] = 0.5*(u[:, 1:]+u[:, :-1])
            Vf[0] = 0.5*(v[1:, :]+v[:-1, :])
            Pf[0] = 0.25*(p[1:, 1:]+p[:-1, :-1]+p[:-1, 1:]+p[1:, :-1])
            counter_save = 0
            for k in tqdm(range(nt-1)):
                compute_ustar(u, divu, meanv, ustar, Uupper)
                compute_vstar(v, divv, meanu, vstar)
                compute_matrix2D(matrix2D, diagon2D)
                p = compute_pressure(matrix2D, diagon2Dx, diagon2Dy, p)
                compute_newu(ustar, p, u)
                compute_newv(vstar, p, v)
                if k*dt > tsavemesh[counter_save+1]:
                    Uf[counter_save+1] = 0.5*(u[:, 1:]+u[:, :-1])
                    Vf[counter_save+1] = 0.5*(v[1:, :]+v[:-1, :])
                    Pf[counter_save+1] = 0.25*(p[1:, 1:]+p[:-1, :-1]+p[:-1, 1:]+p[1:, :-1])
                    counter_save += 1
        except KeyboardInterrupt as error:
            pass
        except Exception as error:
            print("Exited at time: %.2f/%.2f" % (tmin+k*dt, tmax))
            print("Error = ", error)
            raise error
        finally:
            np.savez(filename, xmesh=xmesh, ymesh=ymesh, tmesh=tsavemesh, U = Uf, V = Vf, P = Pf)
    anim = plot_animated_field(tsavemesh, xmesh, ymesh, U=Uf, V=Vf, P=Pf)
    plt.show()
